pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -106,6 +106,13 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 				return err
 			}
 		}
+
+		// if the Pod does not conform to the StatefulSet's nodeSelector, update the Pod's nodeSelector
+		if !nodeSelectorMatches(set, pod) {
```

Comment:
this breaks in the presence of admission that places additional constraints on the pods' nodeSelectors (e.g. https://kubernetes.io/docs/admin/admission-controllers/#podnodeselector)